File: services/client/src/forms/ir/irswap.js
import { sensitivityGroup } from '../shared/sensitivity-fields'

export const irswapForm = {
  name: 'Interest Rate Swap',
  category: 'IR',
  api: 'calculateIRSwap',
  groups: [
    {
      title: 'Trade Details',
      collapsible: true,
      initiallyCollapsed: false,
      fields: [
        {
          key: 'evaluation_date',
          type: 'date',
          label: 'Evaluation Date',
          required: true,
          defaultValue: new Date('2016-02-05'),
          helpText: 'The date at which the swap will be valued'
        },
        {
          key: 'currency',
          type: 'dropdown',
          label: 'Currency',
          required: true,
          defaultValue: 'EUR',
          options: [
            { label: 'USD - US Dollar', value: 'USD' },
            { label: 'EUR - Euro', value: 'EUR' },
            { label: 'GBP - British Pound', value: 'GBP' },
            { label: 'JPY - Japanese Yen', value: 'JPY' },
            { label: 'AUD - Australian Dollar', value: 'AUD' },
            { label: 'CAD - Canadian Dollar', value: 'CAD' },
            { label: 'CHF - Swiss Franc', value: 'CHF' }
          ],
          filter: true,
          helpText: 'The currency in which the swap is denominated'
        },
        {
          key: 'notional',
          type: 'float',
          label: 'Notional Amount',
          required: true,
          defaultValue: 10000000,
          step: 1000,
          minFractionDigits: 2,
          maxFractionDigits: 2,
          helpText: 'The notional amount of the swap',
          validators: [
            (value) => {
              if (value <= 0) {
                return 'Notional amount must be greater than 0'
              }
              return true
            }
          ]
        },
        {
          key: 'start_date',
          type: 'date',
          label: 'Start Date',
          required: true,
          defaultValue: new Date('2023-02-21'),
          helpText: 'The effective date when the swap begins',
          validators: [
            (value, formData) => {
              if (!value || !formData || !formData.evaluation_date) return true
              if (value < formData.evaluation_date) {
                return 'Start date cannot be before evaluation date'
              }
              return true
            }
          ]
        },
        {
          key: 'end_date',
          type: 'date',
          label: 'Maturity Date',
          required: true,
          defaultValue: new Date('2043-02-21'),
          helpText: 'The date when the swap matures',
          validators: [
            (value, formData) => {
              if (!value || !formData || !formData.start_date) return true
              if (value <= formData.start_date) {
                return 'Maturity date must be after start date'
              }
              return true
            }
          ]
        },
        {
          key: 'calendar',
          type: 'dropdown',
          label: 'Business Day Calendar',
          required: true,
          defaultValue: 'TARGET',
          options: [
            { label: 'US - United States', value: 'US' },
            { label: 'TARGET - European Central Bank', value: 'TARGET' },
            { label: 'UK - United Kingdom', value: 'UK' },
            { label: 'JP - Japan', value: 'JP' },
            { label: 'AU - Australia', value: 'AU' },
            { label: 'CA - Canada', value: 'CA' },
            { label: 'CH - Switzerland', value: 'CH' }
          ],
          filter: true,
          helpText: 'The business day calendar used for date adjustments'
        },
        {
          key: 'payment_convention',
          type: 'dropdown',
          label: 'Payment Convention',
          required: true,
          defaultValue: 'MF',
          options: [
            { label: 'Following', value: 'Following' },
            { label: 'Modified Following', value: 'MF' },
            { label: 'Preceding', value: 'Preceding' },
            { label: 'Modified Preceding', value: 'ModifiedPreceding' }
          ],
          filter: true,
          helpText: 'Business day convention for payment dates'
        }
      ]
    },
    {
      title: 'Fixed Leg',
      collapsible: true,
      initiallyCollapsed: false,
      fields: [
        {
          key: 'fixed_payer',
          type: 'boolean',
          label: 'Fixed Payer',
          required: true,
          defaultValue: false,
          helpText: 'True if paying fixed rate, false if receiving fixed rate'
        },
        {
          key: 'fixed_rate',
          type: 'float',
          label: 'Fixed Rate',
          required: true,
          defaultValue: 0.021,
          min: 0,
          max: 1,
          step: 0.0001,
          minFractionDigits: 4,
          maxFractionDigits: 6,
          helpText: 'The fixed rate in decimal form (e.g., 0.021 for 2.1%)',
          validators: [
            (value) => {
              if (value < 0) {
                return 'Fixed rate cannot be negative'
              }
              if (value > 1) {
                return 'Fixed rate cannot exceed 1.0 (100%)'
              }
              return true
            }
          ]
        },
        {
          key: 'fixed_tenor',
          type: 'dropdown',
          label: 'Fixed Leg Payment Frequency',
          required: true,
          defaultValue: '1Y',
          options: [
            { label: 'Annual', value: '1Y' },
            { label: 'Semi-Annual', value: '6M' },
            { label: 'Quarterly', value: '3M' },
            { label: 'Monthly', value: '1M' }
          ],
          filter: true,
          helpText: 'How frequently fixed rate payments are made'
        },
        {
          key: 'fixed_day_counter',
          type: 'dropdown',
          label: 'Fixed Leg Day Count Convention',
          required: true,
          defaultValue: 'A360',
          options: [
            { label: 'Actual/360', value: 'A360' },
            { label: 'Actual/365', value: 'A365' },
            { label: 'Actual/Actual', value: 'ACT/ACT' },
            { label: '30/360', value: '30/360' }
          ],
          filter: true,
          helpText: 'The convention used to calculate the fraction of a year for fixed leg payments'
        }
      ]
    },
    {
      title: 'Floating Leg',
      collapsible: true,
      initiallyCollapsed: false,
      fields: [
        {
          key: 'floating_payer',
          type: 'boolean',
          label: 'Floating Payer',
          required: true,
          defaultValue: true,
          helpText: 'True if paying floating rate, false if receiving floating rate',
          validators: [
            (value, formData) => {
              if (value === formData?.fixed_payer) {
                return 'In a standard IRS, one leg pays and the other receives. Fixed and floating payer flags should be opposite.'
              }
              return true
            }
          ]
        },
        {
          key: 'floating_index',
          type: 'dropdown',
          label: 'Floating Rate Index',
          required: true,
          defaultValue: 'EUR-EURIBOR-6M',
          options: [
            { label: 'USD-LIBOR-3M', value: 'USD-LIBOR-3M' },
            { label: 'USD-LIBOR-6M', value: 'USD-LIBOR-6M' },
            { label: 'USD-SOFR', value: 'USD-SOFR' },
            { label: 'EUR-EURIBOR-3M', value: 'EUR-EURIBOR-3M' },
            { label: 'EUR-EURIBOR-6M', value: 'EUR-EURIBOR-6M' },
            { label: 'EUR-ESTR', value: 'EUR-ESTR' },
            { label: 'GBP-LIBOR-3M', value: 'GBP-LIBOR-3M' },
            { label: 'GBP-LIBOR-6M', value: 'GBP-LIBOR-6M' },
            { label: 'GBP-SONIA', value: 'GBP-SONIA' },
            { label: 'JPY-LIBOR-3M', value: 'JPY-LIBOR-3M' },
            { label: 'JPY-LIBOR-6M', value: 'JPY-LIBOR-6M' },
            { label: 'JPY-TONAR', value: 'JPY-TONAR' },
            { label: 'AUD-BBSW-3M', value: 'AUD-BBSW-3M' },
            { label: 'AUD-BBSW-6M', value: 'AUD-BBSW-6M' },
            { label: 'CAD-CDOR-3M', value: 'CAD-CDOR-3M' },
            { label: 'CHF-LIBOR-3M', value: 'CHF-LIBOR-3M' },
            { label: 'CHF-LIBOR-6M', value: 'CHF-LIBOR-6M' }
          ],
          filter: true,
          helpText: 'The floating rate index used for the floating leg'
        },
        {
          key: 'floating_tenor',
          type: 'dropdown',
          label: 'Floating Leg Payment Frequency',
          required: true,
          defaultValue: '6M',
          options: [
            { label: 'Annual', value: '1Y' },
            { label: 'Semi-Annual', value: '6M' },
            { label: 'Quarterly', value: '3M' },
            { label: 'Monthly', value: '1M' }
          ],
          filter: true,
          helpText: 'How frequently floating rate payments are made'
        },
        {
          key: 'floating_day_counter',
          type: 'dropdown',
          label: 'Floating Leg Day Count Convention',
          required: true,
          defaultValue: 'A360',
          options: [
            { label: 'Actual/360', value: 'A360' },
            { label: 'Actual/365', value: 'A365' },
            { label: 'Actual/Actual', value: 'ACT/ACT' },
            { label: '30/360', value: '30/360' }
          ],
          filter: true,
          helpText: 'The convention used to calculate the fraction of a year for floating leg payments'
        },
        {
          key: 'floating_spread',
          type: 'float',
          label: 'Floating Spread',
          required: true,
          defaultValue: 0.0,
          step: 0.0001,
          minFractionDigits: 4,
          maxFractionDigits: 6,
          helpText: 'Spread over the floating index in decimal form (e.g., 0.001 for 10bp)'
        },
        {
          key: 'is_in_arrears',
          type: 'boolean',
          label: 'Is In Arrears',
          required: true,
          defaultValue: false,
          helpText: 'Whether the fixing is in arrears (fixed at end of period)'
        },
        {
          key: 'fixing_days',
          type: 'number',
          label: 'Fixing Days',
          required: true,
          defaultValue: 2,
          min: 0,
          max: 10,
          helpText: 'Number of business days between rate fixing and payment period start',
          validators: [
            (value) => {
              if (value < 0) {
                return 'Fixing days cannot be negative'
              }
              if (value > 10) {
                return 'Fixing days cannot exceed 10'
              }
              return true
            }
          ]
        }
      ]
    },
    sensitivityGroup
  ]
}

File: services/server/project/qld/ORETrade/ir/irswap.py
"""
Interest Rate Swap Trade Implementation using ORE.

This module provides IRS pricing using the Open Risk Engine (ORE).
"""

import pandas as pd
import project.qld.engine.parser.date as dateParser

from ..ore_base_trade import OREBaseTrade


class IRSwap(OREBaseTrade):
    """
    Interest Rate Swap trade priced using ORE.

    An IRS exchanges fixed rate payments for floating rate payments
    (or vice versa) in the same currency.
    """

    __name__ = "InterestRateSwap"

    def __init__(self, evaluation_date, currency, notional, start_date, end_date,
                 # Fixed leg parameters
                 fixed_payer, fixed_rate, fixed_tenor, fixed_day_counter,
                 # Floating leg parameters
                 floating_payer, floating_index, floating_tenor, floating_day_counter,
                 floating_spread, is_in_arrears, fixing_days,
                 # Common parameters
                 calendar, payment_convention):
        """
        Initialize Interest Rate Swap trade.

        Args:
            evaluation_date: Valuation date
            currency: Currency (e.g., 'EUR', 'USD')
            notional: Notional amount
            start_date: Start date of the swap
            end_date: End date of the swap

            Fixed leg:
            fixed_payer: True if paying fixed, False if receiving fixed
            fixed_rate: Fixed rate (in decimal, e.g., 0.02 for 2%)
            fixed_tenor: Payment frequency for fixed leg (e.g., '1Y')
            fixed_day_counter: Day count convention for fixed leg (e.g., 'A360')

            Floating leg:
            floating_payer: True if paying floating, False if receiving floating
            floating_index: Floating rate index (e.g., 'EUR-EURIBOR-6M')
            floating_tenor: Payment frequency for floating leg (e.g., '6M')
            floating_day_counter: Day count convention for floating leg
            floating_spread: Spread over index (in decimal)
            is_in_arrears: Whether fixing is in arrears
            fixing_days: Number of fixing days

            Common:
            calendar: Business day calendar (e.g., 'TARGET', 'US')
            payment_convention: Payment convention (e.g., 'MF', 'Following')
        """
        super().__init__(evaluation_date)

        self.currency = currency
        self.notional = notional
        self.start_date = start_date
        self.end_date = end_date
        self.fixed_payer = fixed_payer
        self.fixed_rate = fixed_rate
        self.fixed_tenor = fixed_tenor
        self.fixed_day_counter = fixed_day_counter
        self.floating_payer = floating_payer
        self.floating_index = floating_index
        self.floating_tenor = floating_tenor
        self.floating_day_counter = floating_day_counter
        self.floating_spread = floating_spread
        self.is_in_arrears = is_in_arrears
        self.fixing_days = fixing_days
        self.calendar = calendar
        self.payment_convention = payment_convention

        self._parse_input()
        self._validate_input()

    def _parse_input(self):
        """Parse and convert string inputs to appropriate types."""
        self.evaluation_date = dateParser.parse_date(self.evaluation_date)
        self.currency = self.currency.upper()
        self.notional = float(self.notional)
        self.start_date = dateParser.parse_date(self.start_date)
        self.end_date = dateParser.parse_date(self.end_date)
        self.fixed_rate = float(self.fixed_rate)
        self.floating_spread = float(self.floating_spread)
        self.fixing_days = int(self.fixing_days)

        # Parse boolean flags
        if isinstance(self.fixed_payer, str):
            self.fixed_payer = self.fixed_payer.lower() in ('true', 'yes', '1')
        else:
            self.fixed_payer = bool(self.fixed_payer)

        if isinstance(self.floating_payer, str):
            self.floating_payer = self.floating_payer.lower() in ('true', 'yes', '1')
        else:
            self.floating_payer = bool(self.floating_payer)

        if isinstance(self.is_in_arrears, str):
            self.is_in_arrears = self.is_in_arrears.lower() in ('true', 'yes', '1')
        else:
            self.is_in_arrears = bool(self.is_in_arrears)

    def _validate_input(self):
        """Validate trade parameters."""
        if self.notional <= 0:
            raise ValueError("Notional must be positive")

        if self.start_date >= self.end_date:
            raise ValueError("End date must be after start date")

        if self.start_date < self.evaluation_date:
            raise ValueError("Start date cannot be before evaluation date")

        if self.fixing_days < 0:
            raise ValueError("Fixing days must be non-negative")

        # In a standard IRS, one leg pays and the other receives
        if self.fixed_payer == self.floating_payer:
            raise ValueError("Fixed and floating payer flags should be opposite")

    def _build_trade_dataframe(self):
        """
        Build pandas DataFrame with Interest Rate Swap trade data for ORE.

        Returns:
            tuple: ('Swap', pd.DataFrame with trade parameters)
        """
        trade_df = pd.DataFrame({
            'TradeId': [self.__name__],
            'TradeType': [self.__name__],
            'CounterParty': ['CPTY_A'],
            'Currency': [self.currency],
            'Notional': [self.notional],
            'StartDate': [dateParser.ql2pydate(self.start_date).strftime('%Y-%m-%d')],
            'EndDate': [dateParser.ql2pydate(self.end_date).strftime('%Y-%m-%d')],
            'Calendar': [self.calendar],
            'PaymentConvention': [self.payment_convention],

            # Fixed leg
            'FixedPayer': [str(self.fixed_payer).lower()],
            'FixedRate': [self.fixed_rate],
            'FixedTenor': [self.fixed_tenor],
            'FixedDayCounter': [self.fixed_day_counter],

            # Floating leg
            'FloatingPayer': [str(self.floating_payer).lower()],
            'FloatingIndex': [self.floating_index],
            'FloatingTenor': [self.floating_tenor],
            'FloatingDayCounter': [self.floating_day_counter],
            'FloatingSpread': [self.floating_spread],
            'IsInArrears': [str(self.is_in_arrears).lower()],
            'FixingDays': [self.fixing_days]
        })

        return (self.__name__, trade_df)

    def _filter_epe_dates(self, exposure_df):
        """
        Filter exposure DataFrame to include only relevant dates.

        Args:
            exposure_df (pd.DataFrame): DataFrame with exposure data including 'Date' column.
        Returns:
            pd.DataFrame: Filtered DataFrame.
        """
        exposure_df['Date'] = pd.to_datetime(exposure_df['Date'])
        start_date = dateParser.ql2npdate(self.start_date)
        end_date = dateParser.ql2npdate(self.end_date)
        filtered_df = exposure_df[(exposure_df['Date'] >= start_date) & (exposure_df['Date'] <= end_date)]
        return filtered_df
